R/derive_merged.R

```diff
@@ -0,0 +1,685 @@
+    add_data <- filter_extreme(add_data,
+                               by_vars = by_vars,
+                               order = order,
+                               mode = mode,
+                               check_type = check_type)
+  } else {
+    signal_duplicate_records(
+      add_data,
+      by_vars = by_vars,
+      msg = paste(
+        "Dataset `dataset_add` contains duplicate records with respect to",
+        enumerate(vars2chr(by_vars))
+      )
+    )
+  }
+  if (!is.null(new_vars)) {
+    add_data <- select(add_data, !!!by_vars, !!!new_vars)
+  }
+  if (!quo_is_null(match_flag)) {
+    add_data <- mutate(add_data,
+                       !!match_flag := TRUE)
+  }
+  # check if there are any variables in both datasets which are not by vars
+  # in this case an error is issued to avoid renaming of varibles by left_join()
+  common_vars <-
+    setdiff(intersect(names(dataset), names(add_data)), vars2chr(by_vars))
+  if (length(common_vars) > 0L) {
+    abort(if_else(
+      length(common_vars) == 1L,
+      paste0(
+        "The variable ",
+        common_vars[[1]],
+        " is contained in both datasets.\n",
+        "Please add it to `by_vars` or remove or rename it in one of the datasets."
+      ),
+      paste0(
+        "The variables",
+        enumerate(common_vars),
+        "are contained in both datasets.\n",
+        "Please add them to `by_vars` or remove or rename them in one of the datasets."
+      )
+    ))
+  }
+  left_join(dataset, add_data, by = vars2chr(by_vars))
+}
+
+#' Merge a (Imputed) Date Variable
+#'
+#' Merge a imputed date variable and date imputation  flag from a dataset to the
+#' input dataset. The observations to merge can be selected by a condition
+#' and/or selecting the first or last observation for each by group.
+#'
+#' @param dataset_add Additional dataset
+#'
+#'   The variables specified by the `by_vars`, the `dtc`, and the `order`
+#'   parameter are expected.
+#'
+#' @inheritParams derive_vars_merged
+#' @inheritParams derive_vars_dt
+#'
+#' @details
+#'
+#'   1. The additional dataset is restricted to the observations matching the
+#'   `filter_add` condition.
+#'
+#'   1. The date variable and if requested, the date imputation flag is added to
+#'   the additional dataset.
+#'
+#'   1. If `order` is specified, for each by group the first or last observation
+#'   (depending on `mode`) is selected.
+#'
+#'   1. The date and flag variables are merged to the input dataset.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @keywords derivation adam timing
+#'
+#' @export
+#'
+#' @examples
+#' library(admiral.test)
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("dm")
+#' data("ex")
+#'
+#' # derive treatment start date (TRTSDT)
+#' derive_vars_merged_dt(
+#'   select(dm, STUDYID, USUBJID),
+#'   dataset_add = ex,
+#'   by_vars = vars(STUDYID, USUBJID),
+#'   new_vars_prefix = "TRTS",
+#'   dtc = EXSTDTC,
+#'   date_imputation = "first",
+#'   order = vars(TRTSDT),
```

Comment:
It sounds strange to me to use the derived variable as input to the function.
Should it refers only to variable in dataset_add and then inside the function, the derived variable is used in addition to the specified var?
(just thinking)
` order = vars(EXTRT, EXSEQ)` 
and when you call derive_meged_vars you add the derived variable to the order var?: 
`order = vars(EXTRT, EXSEQ, TRTSDT)`